src/Core/Animation/KeyFramedValueInterpolators.hpp

```diff
@@ -0,0 +1,114 @@
+#ifndef RADIUMENGINE_KEYFRAMEDVALUE_INTERPOLATOR_HPP
+#define RADIUMENGINE_KEYFRAMEDVALUE_INTERPOLATOR_HPP
+
+#include <Core/Animation/KeyFramedValue.hpp>
+#include <Core/Animation/Pose.hpp>
+#include <Core/Animation/PoseOperation.hpp>
+#include <Core/Math/Interpolation.hpp>
+
+namespace Ra {
+namespace Core {
+namespace Animation {
+
+/**
+ * Look for the keyframes around time \p t.
+ * \param[in] keyframes The list of keyframes.
+ * \param[in] t The time to search for.
+ * \param[out] t0 The time for the keyframe preceding t.
+ * \param[out] t1 The time for the keyframe following t.
+ * \param[out] dt Linear parameter, \f$ \in [0;1] \f$, of \p t between \p t0 and \p t1.
+ * \note If \p t is lower than the first keyframe's time tf,
+ *       then \p t0 = \p t1 = tf and \p dt = 0.
+ *       If \p t is higher than the last keyframe's time tl,
+ *       then \p t0 = \p t1 = tl and \p dt = 0.
+ *       If \p t is an exact match on a keyframe's time tt,
+ *       then \p t0 = \p t1 = tt and \p dt = 0.
+ */
+template <typename T>
+void findRange( const typename KeyFramedValue<T>::KeyFrames& keyframes,
+                const Scalar& t,
+                Scalar& t0,
+                Scalar& t1,
+                Scalar& dt ) {
+    // before first
+    if ( t < keyframes.begin()->first )
+    {
+        t0 = keyframes.begin()->first;
+        t1 = t0;
+        dt = 0.0;
+        return;
+    }
+    // after last
+    if ( t > keyframes.rbegin()->first )
+    {
+        t0 = keyframes.rbegin()->first;
+        t1 = t0;
+        dt = 0.0;
+        return;
+    }
+    auto it = keyframes.find( t );
+    // exact match
+    if ( it != keyframes.end() )
+    {
+        t0 = it->first;
+        t1 = t0;
+        dt = 0.0;
+        return;
+    }
+    // in-between
+    auto upper = keyframes.upper_bound( t );
```

Comment:
Avoid traversing the whole collection twice: 
```suggestion
    auto upper = std::upper_bound(it, keyframes.end(), t );
```

Reply:
If we don't find `t` in `keyframes` then we will have `it == keyframes.end()`, so your proposal must not work as expected

Reply:
This is case is detected few lines before
https://github.com/hoshiryu/Radium-Engine/blob/keyframedvalue/src/Core/Animation/KeyFramedValueInterpolators.hpp#L51

Reply:
what I meant is that when we come to this line, `it == keyframes.end`, so your proposal doesn't work since it will be the same as 
```std::upper_bound( keyframes.end(), keyframes.end(), t );```

Reply:
what I meant is that when we come to this line, `it == keyframes.end()`, so your proposal doesn't work since it will be the same as  
```std::upper_bound( keyframes.end(), keyframes.end(), t );```